lib/BSON.pm

```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
+
+    use JSON::MaybeXS;
+    my $ext = $bson->perl_to_extjson($data);
+    my $json = encode_json($ext, \%options);
+
+Takes a perl data structure and turns it into an Extended JSON
+structure. Note that the structure will still have to be serialized.
+
+Possible options are:
+
+=for :list
+* C<relaxed> A boolean indicating wether relaxed extended JSON should
+be generated. If not set, the value is defaulted to the C<BSON_EXTJSON>
+environment variable.
+
+=cut
+
+sub perl_to_extjson {
+    my ($class, $data, $options) = @_;
+
+    local $ENV{BSON_EXTJSON} = !$options->{relaxed}
+        if exists $options->{relaxed};
+
+    if (not defined $data) {
+        return undef;
+    }
+
+    if (blessed($data) and $data->can('TO_JSON')) {
+        my $json_data = $data->TO_JSON;
+        return $json_data;
+    }
+
+    if (not ref $data) {
+
+        if (!$ENV{BSON_EXTJSON}) {
+            return $data;
+        }
+        else {
+            if (looks_like_number($data)) {
+                if ($data =~ m{\A-?[0-9_]+\z}) {
+                    if ($data <= $max_int32) {
+                        return { '$numberInt' => "$data" };
+                    }
+                    else {
+                        return { '$numberLong' => "$data" };
+                    }
+                }
+                else {
+                    return { '$numberDouble' => 'Infinity' }
+                        if $data eq 'Inf';
+                    return { '$numberDouble' => '-Infinity' }
+                        if $data eq '-Inf';
+                    return { '$numberDouble' => 'NaN' }
+                        if $data eq 'NaN';
+                    my $value = "$data";
+                    $value = $value / 1.0;
+                    return { '$numberDouble' => "$value" };
+                }
+            }
+
+            return $data;
+        }
+    }
+
+    if (boolean::isBoolean($data)) {
+        return $data;
+    }
+
+    if (ref $data eq 'HASH') {
+        if (exists $data->{'$type'}) {
+            return { '$type' => $class->perl_to_extjson($data->{'$type'}, $options) };
+        }
+        for my $key (keys %$data) {
+            my $value = $data->{$key};
+            $data->{$key} = $class->perl_to_extjson($value, $options);
+        }
+        return $data;
+    }
+    
+    if (ref $data eq 'ARRAY') {
+        for my $index (0 .. $#$data) {
+            my $value = $data->[$index];
+            $data->[$index] = $class->perl_to_extjson($value, $options);
+        }
+        return $data;
+    }
+
+    if (blessed($data) and $data->isa('JSON::PP::Boolean')) {
+        return $data;
+    }
+
+    die sprintf "Unsupported ref value (%s)", ref($data);
+}
+
+=method extjson_to_perl
+
+    use JSON::MaybeXS;
+    my $ext = decode_json($json);
+    my $data = $bson->extjson_to_perl($ext);
+
+Takes an Extended JSON data structure and inflates it into a Perl
+specific data structure. Note that you have to decode the JSON string
+manually beforehand.
```

Comment:
Pls document the conditions under which this will wrap numbers for type preservation.  It looks like it will wrap anything expressed "canonically" as a hashref and won't wrap anything expressed as a scalar. Is that right?